test/test_batchnorm_layer.f90

```diff
@@ -0,0 +1,65 @@
+program test_batchnorm_layer
+
+  use iso_fortran_env, only: stderr => error_unit
+  use nf, only: batchnorm, layer
+  use nf_batchnorm_layer, only: batchnorm_layer
+
+  implicit none
+
+  type(layer) :: bn_layer
+  integer, parameter :: num_features = 64
+  real, allocatable :: sample_input(:,:)
+  real, allocatable :: output(:,:)
+  real, allocatable :: gradient(:,:)
+  integer, parameter :: input_shape(1) = [num_features]
+  real, allocatable :: gamma_grad(:), beta_grad(:)
+  real, parameter :: tolerance = 1e-7
+  logical :: ok = .true.
+
+  bn_layer = batchnorm(num_features)
+
+  if (.not. bn_layer % name == 'batchnorm') then
+    ok = .false.
+    write(stderr, '(a)') 'batchnorm layer has its name set correctly.. failed'
+  end if
+
+  if (bn_layer % initialized) then
+    ok = .false.
+    write(stderr, '(a)') 'batchnorm layer should not be marked as initialized yet.. failed'
+  end if
+
+  ! Initialize sample input and gradient
+  allocate(sample_input(num_features, 1))
+  allocate(gradient(num_features, 1))
+  sample_input = 1.0
+  gradient = 2.0
+
+  !TODO run forward and backward passes directly on the batchnorm_layer instance
```

Comment:
I'll add a simple test directly on the `batchnorm_layer` instance rather than the high-level `layer_type` instance so you get the idea how it will work.